flamingpy/noise/cv.py

```diff
@@ -0,0 +1,623 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Continuous-variable noise model classes."""
+
+# pylint: disable=import-outside-toplevel
+
+import numpy as np
+from numpy.random import default_rng
+
+from flamingpy.codes import EGraph
+from flamingpy.cv.ops import invert_permutation, SCZ_mat, SCZ_apply, splitter_symp
+from flamingpy.cv.gkp import GKP_binner, Z_err_cond
+
+# pylint: disable=too-many-instance-attributes
+class CVLayer:
+    """A class for applying to a code (or an graph state) a physical layer of
+    continuous-variable noise.
+
+    Associates the nodes of an EGraph with continuous-variable quantum states,
+    and its edges with continuous-variable CZ gates.
+
+    Currently, a hybrid lattice of p-squeezed and GKP+ states is considered.
+
+    Args:
+        code (SurfaceCode or EGraph): the code object (so that code.graph is
+            an EGraph), or an EGraph directly, to which the noise is applied.
+        delta (float): the quadrature blurring parameter, related to the
+            squeezing of the GKP states and the momentum-quadrature variance of
+            the p-squeezed states.
+        states (dict, optional): the dictionary of all non-GKP states and their
+            indices, of the form {'state': []}. By default, all states are
+            treated as GKP states.
+        p_swap (float, optional): if supplied, the probability of a node being
+            a p-squeezed state. Overrides the indices given in state.
+        sampling_order (str, optional): the scheme for conducting sampling
+            for the homodyne measurements. Options are "initial" and "two-step".
+        rng (numpy.random.Generator, optional): a random number generator
+            following the NumPy API. It can be seeded for reproducibility.
+            By default, numpy.random.default_rng is used without a fixed seed.
+
+    Attributes:
+        egraph (EGraph): the underlying graph representation.
+        to_points (dict): the index-to-coordinate dictionary, taken from egraph.
+        N (int): the number of qubits in the lattice.
+        delta (float): the delta from the Args above.
+        p_swap (float): the swap-out probability from the Args above.
+        _adj (sp.sparse.csr_matrix): the adjacency matrix of egraph.
+        _states (dict): states along with their indices.
+        _sampling_order (str): sampling order from above.
+        _perfect_inds (list or NoneType): the indices of qubits to treat as
+            ideal.
+    """
+
+    def __init__(self, code, *, delta, **kwargs):
```

Comment:
I have never seen the * used this way. And I can't easily find documentation online. So, I have no idea what it is doing here.